### ISSUE TITLE
Use DaemonSet as part of kube-up.

### ISSUE THREAD

**Issue by erictune** [kubernetes/kubernetes#15324]
We should use DaemonSet to deploy fluentd, and maybe also kube-proxy and kube-registry-proxy.

Issues:
- what creates daemons-sets?  Kube-up.sh or kube-addons.sh?
- upgrading 
- upgrading across an API version change (DaemonSet graduates out of experimental)
- upgrading getting started guides to the new plan.  When to do this, given that we may modify the DaemonSet API based on experience.
- Presumably we start with GCE.   When is it ready for GKE?  
- node upgrades on GKE were separated into two steps.  With this change, 3 steps: update master, update managed node components (DaemonSets), upgrade base node software (user does this).

Benefits:
- deletes some salt config, make all getting-started guides more similar.
- less differences between different getting-started-guides.
- automated update of more software components on GKE

**Comment by erictune** [comment]
I think @mikedanese has done of bunch of thinking in this area.

**Comment by roberthbailey** [comment]
The DaemonSet should be created as a Deployment (the same object we want to transition the addons to). 

We shouldn't do this until DaemonSet is enabled by default (no longer an experimental API).

**Comment by erictune** [comment]
@roberthbailey regarding Deployment: @bgrant0607 said he wants @DaemonSet to evolve to be its own Deployment, and not to use Deployment, IIUC.

@roberthbailey regarding waiting until DaemonSet is enabled by default: In order for DaemonSet to be "no longer experimental", I think we have to have confidence that we aren't going to need to make major changes to it.  I don't think we can get that confidence without using it for realz.  So, I think we should think about the following sequence:

1.  *Alpha DaemonSet Cluster Creation*
  1. Implement use of DaemonSet alongside the existing salt code.  
  1. Use of DaemonSet for kube-up remains optional, controlled by an env var `USE_DAEMON_SETS` which defaults to false.
  1. Just needs to work on GCE at first
  1. start running some GCE e2es with `USE_DAEMON_SETS=1` in parallel with existing e2es.
  1. DaemonSet remains at `apiVersion: experimental: v1alpha1`
1. *Alpha DaemonSet Cluster Upgrades*
  1. Add rolling upgrade features to DaemonSet
  1. run a `kubernetes-gce-upgrade` e2e but with `USE_DAEMON_SETS=1` 
  1.  DaemonSet is still experimental and alpha
1. *Graduate to "beta"*
  1. Move DaemonSet to `experimental/v1beta1` 
  1. Make  `USE_DAEMON_SETS=1` default on GCE.
  1. Start to bring other providers to support  `USE_DAEMON_SETS=1` 
1. *GKE uses  `USE_DAEMON_SETS=1`*
  1.  `USE_DAEMON_SETS=1` on GKE
1. *Cleanup*
  1. warn lagging getting-started-guides
  1. delete old salt code

**Comment by bgrant0607** [comment]
@erictune is correct re. deployment vs. daemonset. We should use Deployment for addons. Also thinking about a role for Deployment Manager.

**Comment by derekwaynecarr** [comment]
this seems like a realistic set of goals

**Comment by paralin** [comment]
+1

**Comment by bgrant0607** [comment]
This is obsolete.

**Comment by roberthbailey** [comment]
Why is this obsolete? We just started switching fluentd to use a daemonset, and we should move kube-proxy to use them as well.
